stan/math/prim/meta/holder.hpp

```diff
@@ -0,0 +1,274 @@
+struct traits<stan::math::Holder<ArgType, Ptrs...>> {
+  typedef typename ArgType::StorageKind StorageKind;
+  typedef typename traits<ArgType>::XprKind XprKind;
+  typedef typename ArgType::StorageIndex StorageIndex;
+  typedef typename ArgType::Scalar Scalar;
+  enum {
+    Flags = (ArgType::Flags
+             & (RowMajorBit | LvalueBit | LinearAccessBit | DirectAccessBit
+                | PacketAccessBit | NoPreferredStorageOrderBit))
+            | NestByRefBit,
+    RowsAtCompileTime = ArgType::RowsAtCompileTime,
+    ColsAtCompileTime = ArgType::ColsAtCompileTime,
+    MaxRowsAtCompileTime = ArgType::MaxRowsAtCompileTime,
+    MaxColsAtCompileTime = ArgType::MaxColsAtCompileTime,
+    InnerStrideAtCompileTime = ArgType::InnerStrideAtCompileTime,
+    OuterStrideAtCompileTime = ArgType::OuterStrideAtCompileTime
+  };
+};
+
+}  // namespace internal
+}  // namespace Eigen
+
+namespace stan {
+namespace math {
+
+/**
+ * A no-op Eigen operation. This object also owns pointers to dynamically
+ * allocated objects used in its argument expression. When this object is
+ * destructed, those objects are deleted.
+ * @tparam Derived derived type
+ * @tparam T type of the argument
+ */
+template <typename ArgType, typename... Ptrs>
+class Holder
+    : public Eigen::internal::dense_xpr_base<Holder<ArgType, Ptrs...>>::type {
+ public:
+  typedef typename Eigen::internal::ref_selector<Holder<ArgType, Ptrs...>>::type
+      Nested;
+  typename Eigen::internal::ref_selector<ArgType>::non_const_type m_arg;
+  std::tuple<std::unique_ptr<Ptrs>...> m_unique_ptrs;
+
+  explicit Holder(ArgType&& arg, Ptrs*... pointers)
+      : m_arg(arg), m_unique_ptrs(std::unique_ptr<Ptrs>(pointers)...) {}
+
+  Holder(const Holder<ArgType, Ptrs...>&) = default;
+  Holder(Holder<ArgType, Ptrs...>&&) = default;
+
+  Eigen::Index rows() const { return m_arg.rows(); }
+  Eigen::Index cols() const { return m_arg.cols(); }
+  Eigen::Index innerStride() const { return m_arg.innerStride(); }
+  Eigen::Index outerStride() const { return m_arg.outerStride(); }
+
+  auto* data() { return m_arg.data(); }
+
+  template <typename T, require_eigen_t<T>* = nullptr>
+  inline Holder<ArgType, Ptrs...>& operator=(const T& other) {
+    m_arg = other;
+    return *this;
+  }
+
+  inline Holder<ArgType, Ptrs...>& operator=(
+      const Holder<ArgType, Ptrs...>& other) {
+    m_arg = other;
+    return *this;
+  }
+
+  inline Holder<ArgType, Ptrs...>& operator=(Holder<ArgType, Ptrs...>&& other) {
+    m_arg = std::move(other.m_arg);
+    return *this;
+  }
+};
+
+}  // namespace math
+}  // namespace stan
+
+namespace Eigen {
+namespace internal {
+
+template <typename ArgType, typename... Ptrs>
+struct evaluator<stan::math::Holder<ArgType, Ptrs...>>
+    : evaluator_base<stan::math::Holder<ArgType, Ptrs...>> {
+  typedef stan::math::Holder<ArgType, Ptrs...> XprType;
+  typedef typename remove_all<typename nested_eval<ArgType, 1>::type>::type
+      ArgTypeNestedCleaned;
```

Comment:
Is there a reason not to have 

```
typedef typename internal::nested_eval<ArgType,Factor>::type ArgTypeNested;
```

Reply:
It was that way in the tutorial I followed. I removed it since it was only used once. What is the reason you would prefere it that way?